### ISSUE TITLE
"Change Dates" uses UTC time, expected MST trips not showing up

### ISSUE THREAD

**Issue by sebastianbarry** [e-mission/e-mission-docs#758]
On the dashboard, there is the "Change Dates" button. When a date range is chosen, the current app looks at those dates/times in UTC (which is about 6 hours later than MST). This causes confusion for users, for example:

Say I want to see how far my run was on Monday 7/18/2022 which I started at 7:00pm MST.
So I go to my app and put in the date range of 7/18/20200 - 7/18/2022 but I don't see my run in the list.
Why is it not in the list? Because 7:00pm MST = 1:00am UTC, 7/19/2022 and I did not input that date range.

So we need to resolve this by aligning the "Change Dates" dates/times with the device's local time

**Comment by shankari** [comment]
The dashboard HTML is at:
https://github.com/e-mission/e-mission-phone/blob/master/www/templates/main-metrics.html

The dashboard javascript is at:
https://github.com/e-mission/e-mission-phone/blob/master/www/js/metrics.js

On the server, we always search by UTC because we want to use it in Colorodo (MT), Durham (ET)...
So we want to change the timezone based on where the user is, and not where the server is.

In general while working with time, you want to work with standard time and convert only at the presentation layer.
otherwise, if we wanted to have a dataset that combined across locations, we would be very confused.

So the library that we use for time is called moment.
https://momentjs.com/

And the change date selector is:
https://github.com/e-mission/e-mission-phone/blob/master/www/templates/metrics/metrics-control.html

As you can see: `datepickerObjFrom.inputMoment.format('LL')` is what we use to show the date.
and so  `format('LL')` should format in the local locale
you should see how we create `inputMoment` and how to change it as needed.

Another change you might want to think about while you are here is what the refresh icon does.
Right now, I think that the refresh icon just re-reads for the same selected date range.
Should it also reset to today? Or should there be a different button for it or ???

**Comment by sebastianbarry** [comment]
In order to test out this feature, I want to be able to put trips into my emulator app, and then attempt changing the code to send the "local version" of the time (-6 hours the current UTC time) to the server for a query

I am attempting to get my emulator app to grab the trips data from my own personal version of the app, however I am struggling in getting this to work. I know there is the [e-mission-server README](https://github.com/e-mission/e-mission-server/blob/master/README.md) which has directions on how to load test data, but this seems slightly different to what I'm trying to do.
_Tyler (paraphrased by me) said the way he accomplished this is by **adding the www/json/connectionConfig.json file**,  **setting his user token to the staging token he uses on his personal app**, and **not running docker**_
```
connectionConfig.json
{
    "connectUrl": "http://stage.canbikeco.org",
    "aggregate_call_auth": "no_auth",
    "android": {
        "auth": {
            "method": "prompted-auth",
            "prompt": "ignored"
        }
    },
    "ios": {
        "auth": {
            "method": "prompted-auth",
            "prompt": "ignored"
        }
    }
}
```

My staging token (used on CanBikeCO): **stage_u6cmj-oVByXvp1JH**

My emulator version of the devapp is using: https://github.com/e-mission/e-mission-devapp/releases/tag/v3.2.4

My emulator version of the devapp is on version: **Study**

My docker is not running

I copied these steps, and my emulator does not get past the login setup page. 

https://user-images.githubusercontent.com/61334340/181642623-01770c84-6326-4172-bfd0-b955a8dab4fe.mov


With docker running, my app does get past the login setup page, but still does not appear to be loading in the test/my own data. I also noticed that my staging token name gets changed in the app, which may be a cause of this:
![Simulator Screen Shot - iPhone 11 - 2022-07-28 at 15 36 41](https://user-images.githubusercontent.com/61334340/181641914-d3642482-d985-4b5c-a6fd-c4ce4da11b77.png)

-------
@shankari Do you have any other document you can guide me to get my data loaded onto my own app
--------

**Comment by shankari** [comment]
@sebastianbarry 

> In order to test out this feature, I want to be able to put trips into my emulator app, and then attempt changing the code to send the "local version" of the time (-6 hours the current UTC time) to the server for a query

Why? Is there a specific reason that you need to use your trips instead of the test data? 
Concretely, if you load test data for 2016 Aug 15th, what prevents you from testing this feature?

Tyler is connecting to the canbikeco server because he is here for a short time, and his feature does not require him to write server code. But I think you should get familiar with end-to-end testing in which the app connects to a dev server so that you can learn how to change the server code as well.

So, unless you can articulate a really strong reason why you must access your own data, I would encourage you to follow the instructions to load test data and connect to the server running in the docker container on your laptop.

**Comment by sebastianbarry** [comment]
> Why? Is there a specific reason that you need to use your trips instead of the test data?
> Concretely, if you load test data for 2016 Aug 15th, what prevents you from testing this feature?

Agreed, I do not need to test with my own data. It would be preferable to be able to have a trip start/end between 12:00am - 6:00am UTC, that way I can see the desync happening between when users *think* a trip is happening (between 6:00pm - 12:00am MST) on one date, but when it gets sent to the server it counts as the following date because of the UTC time difference of +6 hours. _Now, I have not been successful in loading the test data, so I am unsure if the test data contains trips that fall within this time. (Otherwise I see there is the ["Creating fake user data"](https://github.com/e-mission/e-mission-server/blob/master/README.md#loading-test-data) that I can use to simulate data in this time period)_

This is a video of my emulator not having the test data loaded in, despite loading in the test data (terminal output of the loading test data below)

https://user-images.githubusercontent.com/61334340/181808494-c283df2e-da10-4fce-b337-b3b95f5b4856.mp4

Here is my terminal output when loading the july 22 test data into my docker
```
(base) sbarry-33771s:~ sbarry$ cd documents/github/e-mission-server
(base) sbarry-33771s:e-mission-server sbarry$ ./e-mission-py.bash bin/debug/load_timeline_for_day_and_user.py emission/tests/data/real_examples/shankari_2015-07-22 test_july_22
Traceback (most recent call last):
  File "bin/debug/load_timeline_for_day_and_user.py", line 9, in <module>
    import bson.json_util as bju
ModuleNotFoundError: No module named 'bson'
(base) sbarry-33771s:e-mission-server sbarry$ source setup/activate.sh
(emission) sbarry-33771s:e-mission-server sbarry$ ./e-mission-py.bash bin/debug/load_timeline_for_day_and_user.py emission/tests/data/real_examples/shankari_2015-07-22 test_july_22
storage not configured, falling back to sample, default configuration
Connecting to database URL localhost
emission/tests/data/real_examples/shankari_2015-07-22
Loading file emission/tests/data/real_examples/shankari_2015-07-22
After registration, test_july_22 -> f724b72c-3825-4baf-8e22-e464c493c52e
Finished loading 0 entries into the usercache and 1906 entries into the timeseries
(emission) sbarry-33771s:e-mission-server sbarry$ ./e-mission-py.bash bin/debug/load_timeline_for_day_and_user.py emission/tests/data/real_examples/shankari_2015-07-22 test_july_22
storage not configured, falling back to sample, default configuration
Connecting to database URL localhost
emission/tests/data/real_examples/shankari_2015-07-22
Loading file emission/tests/data/real_examples/shankari_2015-07-22
After registration, test_july_22 -> f724b72c-3825-4baf-8e22-e464c493c52e
Finished loading 0 entries into the usercache and 1906 entries into the timeseries
(emission) sbarry-33771s:e-mission-server sbarry$ ./e-mission-py.bash bin/debug/intake_single_user.py -e test_july_22
storage not configured, falling back to sample, default configuration
Connecting to database URL localhost
analysis.debug.conf.json not configured, falling back to sample, default configuration
google maps key not configured, falling back to nominatim
nominatim not configured either, place decoding must happen on the client
expectations.conf.json not configured, falling back to sample, default configuration
ERROR:root:habitica not configured, game functions not supported
Traceback (most recent call last):
  File "/Users/sbarry/Documents/GitHub/e-mission-server/emission/net/ext_service/habitica/proxy.py", line 22, in <module>
    key_file = open('conf/net/ext_service/habitica.json')
FileNotFoundError: [Errno 2] No such file or directory: 'conf/net/ext_service/habitica.json'
2022-07-29T10:53:01.277559-06:00**********UUID f724b72c-3825-4baf-8e22-e464c493c52e: moving to long term**********
2022-07-29T10:53:01.289270-06:00**********UUID f724b72c-3825-4baf-8e22-e464c493c52e: updating incoming user inputs**********
2022-07-29T10:53:01.322511-06:00**********UUID f724b72c-3825-4baf-8e22-e464c493c52e: filter accuracy if needed**********
2022-07-29T10:53:01.338089-06:00**********UUID f724b72c-3825-4baf-8e22-e464c493c52e: segmenting into trips**********
2022-07-29T10:53:01.373063-06:00**********UUID f724b72c-3825-4baf-8e22-e464c493c52e: segmenting into sections**********
2022-07-29T10:53:01.388309-06:00**********UUID f724b72c-3825-4baf-8e22-e464c493c52e: smoothing sections**********
2022-07-29T10:53:01.401149-06:00**********UUID f724b72c-3825-4baf-8e22-e464c493c52e: cleaning and resampling timeline**********
2022-07-29T10:53:01.419835-06:00**********UUID f724b72c-3825-4baf-8e22-e464c493c52e: inferring transportation mode**********
Error while inferring modes, timestamp is unchanged
Traceback (most recent call last):
  File "/Users/sbarry/Documents/GitHub/e-mission-server/emission/analysis/classification/inference/mode/pipeline.py", line 41, in predict_mode
    mip.runPredictionPipeline(user_id, time_query)
  File "/Users/sbarry/Documents/GitHub/e-mission-server/emission/analysis/classification/inference/mode/pipeline.py", line 139, in runPredictionPipeline
    self.loadModelStage()
  File "/Users/sbarry/Documents/GitHub/e-mission-server/emission/analysis/classification/inference/mode/pipeline.py", line 156, in loadModelStage
    self.model = seedp.ModeInferencePipelineMovesFormat.loadModel()
  File "/Users/sbarry/Documents/GitHub/e-mission-server/emission/analysis/classification/inference/mode/seed/pipeline.py", line 93, in loadModel
    fd = open(SAVED_MODEL_FILENAME, "r")
FileNotFoundError: [Errno 2] No such file or directory: 'seed_model.json'
2022-07-29T10:53:01.442760-06:00**********UUID f724b72c-3825-4baf-8e22-e464c493c52e: inferring labels**********
2022-07-29T10:53:01.455005-06:00**********UUID f724b72c-3825-4baf-8e22-e464c493c52e: populating expectations**********
2022-07-29T10:53:01.466544-06:00**********UUID f724b72c-3825-4baf-8e22-e464c493c52e: creating confirmed objects **********
2022-07-29T10:53:01.479873-06:00**********UUID f724b72c-3825-4baf-8e22-e464c493c52e: storing views to cache**********
(emission) sbarry-33771s:e-mission-server sbarry$ 
```

-----
Is there anything in this you can see that I am doing incorrectly to load the test data?
-----

**Comment by shankari** [comment]
All this looks good. The data does seem to have loaded correctly.
Double check the token and the host URL (under Profile -> Developer Zone -> Sync)
Then check the webserver logs (`/var/tmp/webserver.log`) to see what token is received on the server and whether it maps to the correct UUID

**Comment by sebastianbarry** [comment]
My sync URL and token appear to be correct in the Profile screen:
<img width="373" alt="image" src="https://user-images.githubusercontent.com/61334340/181819101-fbe03491-1c49-40f6-8237-0c8009b48473.png">
<img width="372" alt="image" src="https://user-images.githubusercontent.com/61334340/181819028-0556c929-fc36-4c93-95ce-aba709e00060.png">


Here is what I see in `/var/tmp/webserver.log`:
```

2022-07-28 13:52:32,560:DEBUG:123145405378560:Using the skip method to verify id token test_july_22 of length 12
2022-07-28 13:52:32,563:DEBUG:123145405378560:retUUID = f19e23f7-7fbd-4e54-8cee-dd48010e8e15
2022-07-28 13:52:32,566:DEBUG:123145410633728:START POST /datastreams/find_entries/timestamp
2022-07-28 13:52:32,566:DEBUG:123145410633728:methodName = skip, returning <class 'emission.net.auth.skip.SkipMethod'>
2022-07-28 13:52:32,566:DEBUG:123145410633728:Using the skip method to verify id token test_july_22 of length 12
2022-07-28 13:52:32,570:DEBUG:123145410633728:retUUID = f19e23f7-7fbd-4e54-8cee-dd48010e8e15
2022-07-28 13:52:32,661:DEBUG:123145405378560:curr_query = {'user_id': UUID('f19e23f7-7fbd-4e54-8cee-dd48010e8e15'), '$or': [{'metadata.key': 'manual/mode_confirm'}], 'metadata.write_ts': {'$lte': 1659037962, '$gte': -10}}, sort_key = metadata.write_ts
```
Which is the correct token being sent to the webserver as well.
------
This still looks correct, however the data is still not loaded?

**Comment by shankari** [comment]
The sync host is http://stage.canbikeco.org
It should say `http://localhost:8080` if you are connecting to the server on your laptop (localhost is a networking shortcut for the computer you are running on)
stage.canbikeco.org is a server running in the cloud and you don't have access to the logs on it

You should revert the changes that you made by checking with Tyler since you are not connecting the way that he is.

You do see the token in the logs, but they are from yesterday (2022-07-28!!)

You need to remove `connectionConfig.json` and ensure that you are connecting to the server on localhost and it will work.

**Comment by shankari** [comment]
At a high level, the two master branches read the connection configuration very differently.
- For `master`, we read `connectionConfig.json`
- For `master_for_platform`, we read the connection config from the link/QR code that we scan

We should unify these, but I need to figure out how to make it easy for developers to work with it

For now, unless you are working on the intro screens, you should make your changes to master.
Which will work with connectionConfig.json
Further, if you DON'T have a connectionConfig.json and are running from master, we will read the default config for your emulator, which is https://localhost:8080 for iOS. So you should be able to remove it and connect to the server running on your laptop

**Comment by shankari** [comment]
- check the UUID from a successful call (e.g. label screen refresh)
- if that is correct
   - check the timestamps that we are querying on the server

**Comment by shankari** [comment]
translate timestamp into string

```
new Date(ts * 1000)
```

```
import arrow
arrow.get(ts)
```

**Comment by shankari** [comment]
the OpenPATH pipeline generates different types of objects from the different pipestages (see Gabe's paper or my thesis or talk to Gabe for more details).
after the cleaning stage, it generates `analysis/cleaned_section`
after the inference stage, it generates `analysis/inferred_section`
after the user input matching stage, it generates `analysis/confirmed_trip`

When you ran your pipeline, you did not have model saved, and so you got an error

```
2022-07-29T10:53:01.419835-06:00**********UUID f724b72c-3825-4baf-8e22-e464c493c52e: inferring transportation mode**********
Error while inferring modes, timestamp is unchanged
Traceback (most recent call last):
  File "/Users/sbarry/Documents/GitHub/e-mission-server/emission/analysis/classification/inference/mode/pipeline.py", line 41, in predict_mode
    mip.runPredictionPipeline(user_id, time_query)
  File "/Users/sbarry/Documents/GitHub/e-mission-server/emission/analysis/classification/inference/mode/pipeline.py", line 139, in runPredictionPipeline
    self.loadModelStage()
  File "/Users/sbarry/Documents/GitHub/e-mission-server/emission/analysis/classification/inference/mode/pipeline.py", line 156, in loadModelStage
    self.model = seedp.ModeInferencePipelineMovesFormat.loadModel()
  File "/Users/sbarry/Documents/GitHub/e-mission-server/emission/analysis/classification/inference/mode/seed/pipeline.py", line 93, in loadModel
    fd = open(SAVED_MODEL_FILENAME, "r")
FileNotFoundError: [Errno 2] No such file or directory: 'seed_model.json'
```

So when the server is looking for inferred sections, it does not find any.
Fortunately, we can configure which kinds of objects to return in `conf/analysis/debug.conf.sample`

If you change the output to `analysis/cleaned_section` you will get results based on the sensed outputs
If you change the output to `analysis/confirmed_trip`, you will get a range of results based on the user labels

Either will work for this bug :)

**Comment by sebastianbarry** [comment]
I went ahead and changed the output to `analysis/cleaned_section` in `conf/analysis/debug.conf.sample` and changed the filename to `conf/analysis/debug.conf.json` but the app still appears to not be loading the values in the dashboard:

https://user-images.githubusercontent.com/61334340/182465434-b58f7c4c-a231-4dca-9fab-3ebb302c9deb.mov

Am I doing the config file correctly?

**Comment by sebastianbarry** [comment]
I am also stuck on figuring out where the function that **sends the UTC timestamps to the server** is located in `www/js/metrics.js`, 

I see that there is `getUserMetricsFromServer` and `getAggMetricsFromServer`
```javascript
var getUserMetricsFromServer = function() {
      var clonedData = angular.copy(data);
      delete clonedData.metric;
      clonedData.metric_list = METRIC_LIST;
      clonedData.is_return_aggregate = false;
      var getMetricsResult = CommHelper.getMetrics(theMode, clonedData);
      console.log("getMetricsResult -> ", getMetricsResult);
      return getMetricsResult;
   }
   var getAggMetricsFromServer = function() {
      var clonedData = angular.copy(data);
      delete clonedData.metric;
      clonedData.metric_list = METRIC_LIST;
      clonedData.is_return_aggregate = true;
      var getMetricsResult = CommHelper.getAggregateData(
        "result/metrics/timestamp", clonedData)
      return getMetricsResult;
   }
```
^^^ so I placed a console.log to try to understand what is being passed here but I only get
<img width="240" alt="image" src="https://user-images.githubusercontent.com/61334340/182478921-8c3acce6-871a-4abc-805e-e6402f994948.png">

-----

I also placed a console.log here
```javascript
ar setMetrics = function(mode, callback) {
      theMode = mode;
      if (['local_date', 'timestamp'].indexOf(mode) == -1) {
        console.log('Illegal time_type');
        return;
      }

      if (mode === 'local_date') { // local_date filter
        var tempFrom = $scope.selectCtrl.fromDateLocalDate;
        tempFrom.weekday = $scope.selectCtrl.fromDateWeekdayValue;
        var tempTo = $scope.selectCtrl.toDateLocalDate;
        tempTo.weekday = $scope.selectCtrl.toDateWeekdayValue;
        data = {
          freq: $scope.selectCtrl.freq,
          start_time: tempFrom,
          end_time: tempTo,
          metric: ""
        };
      } else if (mode === 'timestamp') { // timestamp range
        if(lastTwoWeeksQuery) {
          var tempFrom = moment2Timestamp(moment().utc().startOf('day').subtract(14, 'days'));
          var tempTo = moment2Timestamp(moment().utc().startOf('day').subtract(1, 'days'))
          lastTwoWeeksQuery = false; // Only get last week's data once
          $scope.defaultTwoWeekUserCall = true;
        } else {
          var tempFrom = moment2Timestamp($scope.selectCtrl.fromDateTimestamp);
          var tempTo = moment2Timestamp($scope.selectCtrl.toDateTimestamp);
          console.log(tempFrom, " -[Timestamp From -> To]-  ", tempTo); // HERE // HERE <*-*-*-*-*-*-*-*-*-*-*-*-*-
          $scope.defaultTwoWeekUserCall = false;
        }
        data = {
          freq: $scope.selectCtrl.pandaFreq,
          start_time: tempFrom,
          end_time: tempTo,
          metric: ""
        };
      } else {
        console.log('Illegal mode');
        return;
      }
      console.log("Sending data "+JSON.stringify(data));
      callback()
    };
```
To attempt to see what the timestamp is being passed during the `setMetrics` function. I see that this one in particular is the correct MST timestamp, not a UTC timestamp (I translated the timestamp using an online Unix timestamp translator):
<img width="655" alt="image" src="https://user-images.githubusercontent.com/61334340/182479402-8e750d8c-8873-4b7f-ae39-bd6208f625e0.png">

### So now I am wondering, is the issue I'm trying to solve, even an issue?

_Now, I'm still having a hard time getting my data to work in the Dashboard (see previous comment on this issue), so I cannot confirm/deny this right now, but it is a fair question ("Can we actually see this behavior happening? Of trips around 6pm MST not counting because of the UTC-timestamp-upload-process to the server?")_

**Comment by shankari** [comment]
> I went ahead and changed the output to analysis/cleaned_section in conf/analysis/debug.conf.sample and changed the filename to conf/analysis/debug.conf.json but the app still appears to not be loading the values in the dashboard:

You need to restart the server. It doesn't auto-reload like the UI.

**Comment by shankari** [comment]
> Of trips around 6pm MST not counting because of the UTC-timestamp-upload-process to the server?")

I don't think this is the issue, since Jeanne was trying to look at a trip from a month ago.

> To attempt to see what the timestamp is being passed during the setMetrics function. I see that this one in particular is the correct MST timestamp, not a UTC timestamp (I translated the timestamp using an online Unix timestamp translator):

Hm, let's debug this further in our meeting today.

**Comment by sebastianbarry** [comment]
I created a Pull Request that has fixed this issue:
https://github.com/e-mission/e-mission-phone/pull/869

----
### Testing

https://user-images.githubusercontent.com/61334340/182939173-c7175ea1-90bc-4e86-acf2-cf5a94fa0a3c.mov

----
### Problem

I noticed that, when you select (for example) July 22 as the 'to' day, the 'to' timestamp gets set to July 22 @ 11:59pm, however the app displays this as being July 23, which may confuse users if they are paying attention, and could cause an issue

**Comment by sebastianbarry** [comment]
I fixed the above problem in PR #869 https://github.com/e-mission/e-mission-phone/pull/869

I also fixed a problem I noticed where a user could input the 'from' day as _after_ the 'to' day, and vice versa

-------

Here is the date displaying properly:
![Simulator Screen Shot - iPhone 11 - 2022-08-08 at 16 13 02](https://user-images.githubusercontent.com/61334340/183524912-86f6f560-e8ec-4167-a07d-97816a2068ac.png)
This was happening because the date was being switched to UTC, so it would be changing the date you picked (july 22 @ 23:59:59 MST) to UTC time (july 23 @ 5:59:59 UTC). I removed the UTC switch and that alleviated the problem

Here is the date range problem working now:

https://user-images.githubusercontent.com/61334340/183525177-2942012b-4bd6-4611-acad-bbe399110d7b.mov

I changed it to check if the inputted date was outside of the range, and then it automatically sets the picked date to within the range (if I picked 'from' as july 22, and 'to' as july 1, it sets the 'to' as july 22, and vice versa)
